subprojects/core/src/main/java/org/gradle/api/internal/plugins/SoftwareTypeRegistrationPluginTarget.java

```diff
@@ -81,23 +87,56 @@ private void registerSoftwareTypes(TypeMetadata typeMetadata) {
-                validateSoftwareTypePluginExposesSoftwareTypes(softwareTypeImplClass);
+                validateSoftwareTypePluginExposesExactlyOneSoftwareType(softwareTypeImplClass, typeMetadata.getType());
                 softwareTypeRegistry.register(softwareTypeImplClass);
             }
         });
     }
 
-    void validateSoftwareTypePluginExposesSoftwareTypes(Class<? extends Plugin<Project>> softwareTypePluginImplClass) {
+    void validateSoftwareTypePluginExposesExactlyOneSoftwareType(Class<? extends Plugin<Project>> softwareTypePluginImplClass, Class<?> registeringPlugin) {
+        DefaultTypeValidationContext typeValidationContext = DefaultTypeValidationContext.withRootType(softwareTypePluginImplClass, false);
         TypeToken<?> softwareTypePluginImplType = TypeToken.of(softwareTypePluginImplClass);
         TypeMetadata softwareTypePluginImplMetadata = inspectionScheme.getMetadataStore().getTypeMetadata(softwareTypePluginImplType.getRawType());
+        softwareTypePluginImplMetadata.visitValidationFailures(null, typeValidationContext);
 
-        for (PropertyMetadata propertyMetadata : softwareTypePluginImplMetadata.getPropertiesMetadata()) {
-            Optional<SoftwareType> registersSoftwareType = propertyMetadata.getAnnotation(SoftwareType.class);
-            if (registersSoftwareType.isPresent()) {
-                return;
-            }
+        long exposedSoftwareTypes = softwareTypePluginImplMetadata.getPropertiesMetadata().stream()
+            .map(propertyMetadata -> propertyMetadata.getAnnotation(SoftwareType.class))
+            .filter(Optional::isPresent)
+            .count();
+
+        if (exposedSoftwareTypes == 0) {
+            typeValidationContext.visitTypeProblem(problem ->
+                problem.withAnnotationType(softwareTypePluginImplClass)
+                    .id("missing-software-type", "Missing software type annotation", GradleCoreProblemGroup.validation().type())
+                    .contextualLabel("is registered as a software type plugin but does not expose a software type")
+                    .severity(Severity.ERROR)
+                    .details("This class was registered as a software type plugin, but it does not expose a software type. Software type plugins must expose exactly one software type via a property with the @SoftwareType annotation.")
+                    .solution("Add @SoftwareType annotations to properties of " + softwareTypePluginImplClass.getSimpleName())
+                    .solution("Remove " + softwareTypePluginImplClass.getSimpleName() + " from the @RegistersSoftwareTypes annotation on " + registeringPlugin.getSimpleName())
+            );
+        } else if (exposedSoftwareTypes > 1) {
+            typeValidationContext.visitTypeProblem(problem ->
+                problem.withAnnotationType(softwareTypePluginImplClass)
+                    .id("missing-software-type", "Missing software type annotation", GradleCoreProblemGroup.validation().type())
```

Comment:
❕ I think this needs a different id - a type is not missing here, so this error should be distinct from the missing case.

Reply:
Good catch.  Copy-and-paste error.